faker/generator.py

```diff
@@ -99,10 +98,24 @@ def parse(self, text):
         """
         Replaces tokens (like '{{ tokenName }}' or '{{tokenName}}')
         with the result from the token method call.
+
+        Parameters can be supplied using a colon, and then key=value
+        pars. For example:
+
+        '{{ color:hue=(100,200) }} - {{ pyint:min_value=1, min_value=10 }}'
         """
         return _re_token.sub(self.__format_token, text)
 
     def __format_token(self, matches):
         formatter = list(matches.groups())
-        formatter[1] = str(self.format(formatter[1]))
-        return ''.join(formatter)
+
+        if formatter[1]:
+            try:
+                params = eval(formatter[1].replace(':', 'dict(') + ')')
```

Comment:
I'd rather tokenize the input, parse it, and return a `dict` instead of  just wrap it in a `dict(..)` sting and pass it to eval.

We could use something like [`sly`](https://github.com/dabeaz/sly).

I've actually done something very similar recently for a different project, but in Faker's case I'm wary of the performance impact of a more complex parsing, especially when methods are called in loops. I'd definitely do some profiling.

Reply:
JSON maybe simpler than Sly, consider we just need a dictionary and it would handle all the data structures need to pass to the providers.  We could pre-parse the JSON to make basic syntax allowed.

Reply:
I agree that JSON would be simpler to implement, but doing string replacement feels hacky, and I fear that a) it would be more fragile (ie: not handling all cases) and b) in order to handle all cases, it could end up be just as complex as sly, if not more.

Consider having to support the following cases:
```
{{ pyint:min_value=1, min_value=10 }}
{{pyint:min_value=1, min_value=10}}
{{ pyint:min_value=1,   min_value=10 }}
{{ pyint:min_value = 1, min_value = 10 }}
{{ pyint : min_value=1, min_value=10 }}

{{ color:hue=[100, 200] }}
{{ color:hue=(100, 200) }}
```

I think it wold helpful to see what the pre-parsing code you have in mind actually looks like

Reply:
You are right about the performance.  It took a big hit so I've put in format caching if parameters are used, and the results were great.  The parser seems to cover all the edge cases listed.   It's quite bullet proof, and even mixed/bad formatting gets corrected, so much so I had to remove my old test for bad formatting.   More tests are needed to ensure __format_json_parser retains stability though and I haven't done it yet.